File: Programmers/Lv3/Lv3_이중우선순위큐.cpp
#include <string>
#include <vector>
#include <list>
#include <sstream>
using namespace std;

vector<int> solution77(vector<string> operations) {
	vector<int> answer;
	list<int> l;
	
	for (int i = 0; i < operations.size(); i++) {
		stringstream split_string(operations[i]);
		string oper;
		string num;
		int real_num = 0;
		split_string >> oper;
		split_string >> num;
		if (oper == "I") {
			if (num[0] == '-')
				real_num -= atoi(num.substr(1, num.length()).c_str());
			else
				real_num += atoi(num.c_str());
			l.push_back(real_num);
			l.sort();
		}
		else {
			if (num == "1") {			// ÃÖ´ñ°ª »èÁ¦ 
				if(l.size()>0)
					l.pop_back();
			}
			else {						// ÃÖ¼Ú°ª »èÁ¦ 
				if (l.size() > 0)
					l.pop_front();
			}
		}
	}

	if (l.size()==0) {
		answer.push_back(0);
		answer.push_back(0);
	}
	else {
		answer.push_back(l.back());
		answer.push_back(l.front());
	}
		
	return answer;
}

File: Programmers/Lv3/Lv3_이중우선순위큐.py
def solution(operations):
    num = []
    for i in operations:
        temp = i.split()
        if temp[0] == 'I':
            num.append(int(temp[1]))
        else:
            if num:
                if temp[1] == '1':
                    num.sort()
                    num.pop()
                else:
                    num.sort()
                    num.pop(0)
    num.sort()
    if num:
        return [num[-1], num[0]]
    else:
        return [0,0]


print(solution(["I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333"]))
# [7,5]